sktime/clustering/spatio_temporal/_st_dbscan.py

```diff
@@ -0,0 +1,296 @@
+        "capability:multivariate": False,
+        "capability:unequal_length": False,
+        "capability:missing_values": False,
+        "capability:multithreading": False,
+        "capability:predict": True,  
+        "capability:predict_proba": False, 
+        "capability:out_of_sample": True, 
+    }
+
+    def __init__(self, eps1=0.5, eps2=10, min_samples=5, metric='euclidean',n_jobs=-1):
+        self.eps = eps1
+        self.eps1 = eps1
+        self.eps2 = eps2
+        self.min_samples = min_samples
+        self.metric = metric
+        self.n_jobs = n_jobs
+
+        super().__init__()
+
+    def _fit(self, X):
+        """Apply the ST DBSCAN algorithm 
+        ----------
+        X : 2D numpy array with
+            The first element of the array should be the time 
+            attribute as float. The following positions in the array are 
+            treated as spatial coordinates. The structure should look like this [[time_step1, x, y], [time_step2, x, y]..]
+            For example 2D dataset:
+            array([[0,0.45,0.43],
+            [0,0.54,0.34],...])
+        Returns
+        -------
+        self
+        """
+
+        X = check_array(X)
+
+        if not self.eps1 > 0.0 or not self.eps2 > 0.0 or not self.min_samples > 0.0:
+            raise ValueError('eps1, eps2, minPts must be positive')
+
+        n, m = X.shape
+
+        if len(X) < 20000:
+            time_dist = pdist(X[:, 0].reshape(n, 1), metric=self.metric)
+            euc_dist = pdist(X[:, 1:], metric=self.metric)
+
+            dist = np.where(time_dist <= self.eps2, euc_dist, 2 * self.eps1)
+
+            db = DBSCAN(eps=self.eps1,
+                        min_samples=self.min_samples,
+                        metric='precomputed')
+            db.fit(squareform(dist))
+
+            self.labels = db.labels_
+
+        else:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+
+                nn_spatial = NearestNeighbors(metric=self.metric,
+                                              radius=self.eps1)
+                nn_spatial.fit(X[:, 1:])
+                euc_sp = nn_spatial.radius_neighbors_graph(X[:, 1:],
+                                                           mode='distance')
+
+                nn_time = NearestNeighbors(metric=self.metric,
+                                           radius=self.eps2)
+                nn_time.fit(X[:, 0].reshape(n, 1))
+                time_sp = nn_time.radius_neighbors_graph(X[:, 0].reshape(n, 1),
+                                                         mode='distance')
+
+                row = time_sp.nonzero()[0]
+                column = time_sp.nonzero()[1]
+                v = np.array(euc_sp[row, column])[0]
+
+                dist_sp = coo_matrix((v, (row, column)), shape=(n, n))
+                dist_sp = dist_sp.tocsc()
+                dist_sp.eliminate_zeros()
+
+                db = DBSCAN(eps=self.eps1,
```

Comment:
I believe there may be other clustering algorithms here (e.g. `OPTICS`). I wonder if the spatial clustering method could be passed as a parameter 🤔 Otherwise `st_optict`, `st_hdbscan`, etc. will have a lot of duplicated code

Reply:
Yeah , we can do this . I have only added code for one algo others are remaining

Reply:
excellent thought! Although this would make most parameteres obsolete, in similar cases where there was a very popular choice and a modular option, we added two estimators, one with hard-coded component, one with a modular component.